src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHandlerManager.cs

```diff
@@ -175,12 +192,30 @@ private void ProcessItems(IComparable version, IProjectChangeDescription project
 
             foreach (ICommandLineHandler handler in handlers)
             {
-                handler.Handle(version, addedItems, removedItems, isActiveContext, logger);
+                EnqueueWork(_ => handler.Handle(version, addedItems, removedItems, isActiveContext, logger));
             }
 
             CommandLineNotifications.FirstOrDefault()?.Value.Invoke(context.BinOutputPath, addedItems, removedItems);
         }
 
+        private void EnqueueWork(Action<Task> action)
+        {
+            if (LanguageServiceHost.WorkspaceSupportsBatchingAndFreeThreadedInitialization())
+            {
+                _currentHandlerTask = _currentHandlerTask.ContinueWith(
```

Comment:
Why form a queue here vs. just let the calls still go on? Given there's a lack of locking on _currentHandlerTask itself clearly this is already being serialized somewhere higher up.

Reply:
(unless you're implementing the "let's let this run later" proposal, but not sure if it's better to keep behavior more similar for now?)

Reply:
I was using this to serialize, but maybe it's better just to lock while making the calls.  The purpose was to have just the events related to the current batch happen within the batch, but I think a better approach for that might be to hold a lock for the length of the batch.

Reply:
These are actually already serialized in LanguageServiceHost.HandleAsync.

Reply:
(so I removed the queuing entirely).